Add files via upload

kr-matthews opened the pull request:
Added ordering, and a bunch of stuff

kr-matthews committed:
Add files via upload

Added ordering, and a bunch of stuff

weixifan commented on Card.java:
```diff
@@ -1,16 +1,37 @@
+import java.util.Comparator;
+
+class CardComparator implements Comparator<Card> {
```
It might be easier to directly make Card itself implement Comparable. It tends to be the case that Java standard libraries accept Comparables whenever Comparators are accepted. The two are very similar.

weixifan commented on Card.java:
```diff
@@ -1,16 +1,37 @@
+    if (c1.getSuit().equals(c2.getSuit())) {
+      return new RankComparator().compare(c1.getRank(), c2.getRank());
+    } else {
+      return new SuitComparator().compare(c1.getSuit(), c2.getSuit());
```
In general,

`if (...) {
  /* stuff X */
  return ...;
} else {
  /* stuff Y */
}`

can be simplified to

`if (...) {
  /* stuff X */
  return ...;
}
/* stuff Y */`

Also, another way to write lexicographical comparisons is to first perform the `SuitComparator` comparison, and return it if it is nonzero, and otherwise return the `RankComparator` comparison result. This way, you avoid calling `.equals()` on the suit when you may as well just do a tiny bit more computation to get the full suit comparison.

kr-matthews replied:
```suggestion
  @Override
  public int compareTo(Card card) {
    int suitOrder = this.getSuit().compareTo(card.getSuit());
    if (suitOrder == 0) {
      return this.getRank().compareTo(card.getRank());
    }
    return suitOrder;
  }
```

kr-matthews replied:
I think I did the suggested change thing wrong, but that's the code I have on my laptop now

weixifan commented on Card.java:
```diff
@@ -1,16 +1,37 @@
   @Override
   public String toString() {
-    return rank + " of " + suit;
+    return rank.toString() + suit.toString();
```
Formatting would be nice:

`return String.format("rank [%s] suit [%s]", rank, suit);`

kr-matthews replied:
I just want to print 2 characters for display purposes. Like when displaying the user's hand.

weixifan commented on Game.java:
```diff
@@ -1,19 +1,37 @@
   private boolean gameOver = false;
 
+  // position i has the cumulative scores of each player up to the end of ith
+  // round
```
Consider making the `RoundResult` object, which is just a holder for 4 integers. This indirection makes the type easier to read (`List<RoundResult>`), and should include a 4-argument constructor that essentially guarantees that it will always be filled with exactly four scores, one from each player. It also abstracts away the round into its own black box, which increases isolation.

In fact, you can even have `CumulativeHistory` be its own object, offering various services like `int getNumberOfRounds()` or `RoundResult getCumulativeScores(int roundNumber)`. The method `addToGameScoresHistory` below would actually be better located as part of `RoundHistory` as a `void addRound(RoundResult roundResult)`. Now `Game` is an isolated concept from the management of cumulative scores, which is nice.

kr-matthews replied:
I don't have a good sense of when to make a new class, as opposed to just an instance variable in another class.

weixifan commented on Rank.java:
```diff
@@ -0,0 +1,32 @@
+
+}
+
+public enum Rank {
```
Can make this implement `Comparable`, so that the `Comparator` above isn't needed.

kr-matthews replied:
Apparently enums already implement Comparable, with the ordinal ordering.

weixifan commented on Rank.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.Comparator;
+
+class RankComparator implements Comparator<Rank> {
```
I didn't know you could have more than one class per file haha.

weixifan commented on Suit.java:
```diff
@@ -1,3 +1,25 @@
+  }
+
+}
+
```
Same here. Can implement `Comparable` and get rid of `SuitComparator` above.

weixifan commented on Card.java:
```diff
@@ -1,16 +1,37 @@
+    return rank.toString() + suit.toString();
   }
 
   @Override
```
There is an extremely important thing you need to do:

- Whenever you override `equals()`, it is strongly encouraged that you also override `hashCode()` so that it remains conformant to its [contract](https://docs.oracle.com/javase/10/docs/api/java/lang/Object.html#hashCode()).

There is a very handy utility class in the Standard Java Library, `java.util.Objects`, which offers the method [hash](https://docs.oracle.com/javase/10/docs/api/java/util/Objects.html#hash(java.lang.Object...)), that you may find very useful in implementing this method.

weixifan reviewed (commented):
First round of reviews. I will review your `Round` logic in a later review.

kr-matthews committed:
Add files via upload

Made changes based on most suggestions given. Still working on some other suggestions.

kr-matthews committed:
Add files via upload

Made a lot of structural changes

kr-matthews committed:
Add files via upload

more minor changes

kr-matthews committed:
Add files via upload

more progress, added Trick class, reorganized Game and Round classes

kr-matthews committed:
Add files via upload

kr-matthews committed:
Add files via upload

more minor updates

kr-matthews committed:
Add files via upload

It actually plays a game to the end now.
Though each player automatically plays the first card in their hand and the player who had the 2 of clubs wins every trick that round (hence shoots the moon).

kr-matthews committed:
Add files via upload

Now correctly picks who won the trick, assigns that player the points, and makes them start the next trick.
Each player still plays the first card in their (sorted) hand, which may not be a valid play.
Still need to make sure shooting the moon doesn't lose you the game, get player name(s), improve formatting of output text, give computer players a strategy, and allow user to select card to play.

kr-matthews committed:
Add files via upload

kr-matthews committed:
Add files via upload

You can pick which card you play.
Organization probably needs improvement.
User input is probably terribly done.
Still doesn't pass 3 cards or accept names. Formatting also still bad. Will let you lose by shooting the moon still.

kr-matthews committed:
Add files via upload

Updated winning condition.
Other display changes.

kr-matthews committed:
Add files via upload

Computer now plays random (valid) card instead of right-most valid card.
